src/printer.js

```diff
@@ -173,26 +263,92 @@ function printMemberChain(path, options, print) {
   //     ->b->c()
   //     ->d()
   const groups = [];
-  let currentGroup = [];
-  let hasSeenCall = false;
-  printedNodes.forEach(printedNode => {
-    if (hasSeenCall && printedNode.node.kind === "propertylookup") {
-      groups.push(currentGroup);
-      currentGroup = [printedNode];
-      hasSeenCall = false;
+  let currentGroup = [printedNodes[0]];
+  let i = 1;
+  for (; i < printedNodes.length; ++i) {
+    if (
+      printedNodes[i].node.kind === "call" ||
+      (printedNodes[i].node.kind === "offsetlookup" &&
+        printedNodes[i].node.offset &&
+        printedNodes[i].node.offset.kind === "number")
+    ) {
+      currentGroup.push(printedNodes[i]);
     } else {
-      currentGroup.push(printedNode);
+      break;
     }
-
-    if (printedNode.node.kind === "call") {
-      hasSeenCall = true;
+  }
+  if (printedNodes[0].node.kind !== "call") {
+    for (; i + 1 < printedNodes.length; ++i) {
```

Comment:
forEach? the iterator here is confusing

Reply:
@mgrip `forEach` is not optimize for this situation, we don't start with `0` value and it is create more memory and cpu usage, also i do not see any problems with reading this

Reply:
I thought `forEach(callback(item, index) {})` just uses the index, which would start at `0`? you're not resetting `i` from the previous loop, wouldn't that cause a problem?

Reply:
@mgrip it is expected behaviour, we break grouping when it is not call (i.e. `$a[1][2][3]->property` should be in one line, reseting `i` break property on newline and it is ugly).